lib/core/utils/filter-html-attrs.js

```diff
@@ -0,0 +1,138 @@
+  `|'([^']*)'` + // group 3: single quoted attribute value; look for everything surrounded by single quotes
+  `|([^'"\\s]*)))` + // group 4: non-quoted attribute value; look for any non-whitespace and quote characters
+    `|\\b(\\S+)(?:\\B|\\b)`, // group 5: non-valued attribute name; look for any remaining non-matched strings surrounded by word boundaries
+  'g'
+);
+
+const attrMap = {
+  for: 'htmlFor',
+  class: 'className'
+};
+
+/**
+ * Match an attribute to a matcher object.
+ * @param {String} nodeName - Name of the current node
+ * @param {Object[]} attributes - Array of node attributes and values
+ * @param {String} value - Value of the current attribute
+ * @param {any} matcher - How to match the value
+ */
+function matches(nodeName, attributes, value, matcher) {
+  switch (typeof matcher) {
+    case 'boolean':
+      return true;
+    case 'string':
+      return matcher === value;
+    case 'object':
+      if (Array.isArray(matcher)) {
+        return matcher.includes(value);
+      }
+
+      // matcher-like object, match all parts
+      let filter = true;
+
+      // match each property
+      if (matcher.nodeName) {
+        filter = filter && matcher.nodeName === nodeName;
+      }
+
+      if (matcher.attributes) {
+        Object.keys(matcher.attributes).forEach(attrName => {
+          const attr = attributes.find(attr => attr.name === attrName);
+          if (!attr) {
+            return false;
+          }
+
+          filter =
+            filter &&
+            matches(
+              nodeName,
+              attributes,
+              attr.value,
+              matcher.attributes[attrName]
+            );
+        });
+      }
+
+      return filter;
+  }
+}
+
+/**
+ * Filter attributes from an html string. This is not to prevent XSS attacks but instead is used to remove attributes which shouldn't appear in the output. As such, we can assume mostly normal attribute output and not attributes purposefully trying to break a parser
+ *
+ * Example:
+ * ```js
+ * // Remove attribute if present regardless of value
+ * axe.utils.filterHtmlAttrs('<div data-attr="foo">my div</div>', { 'data-attr': true });
+ *
+ * // Remove attribute if value matches
+ * axe.utils.filterHtmlAttrs('<div data-attr="foo">my div</div>', { 'data-attr': 'foo' });
+ *
+ * // Remove attribute if value matches list of values
+ * axe.utils.filterHtmlAttrs('<div data-attr="foo">my div</div>', { 'data-attr': ['foo', 'bar'] });
+ *
+ * // Remove attribute if tag matches matcher-like object
+ * axe.utils.filterHtmlAttrs('<div class="foo"><input type="text" class="foo"/></div>', { class: { nodeName: input } });
+ * ```
+ *
+ * @method getRootNode
+ * @memberof axe.utils
+ * @param {String} htmlString - HTML string to remove attributes from.
+ * @param {Object} filterAttrs - Attributes to remove and the qualifier of when to remove them.
+ * @returns {String}
+ */
+function filterHtmlAttrs(htmlString, filterAttrs) {
```

Comment:
I think you should isolate the regex code into its own file / methods, abstract them away, so that the regex logic isn't mixed in with the actual filtering. I think this function should look something like this:

```js
function filterHtmlAttrs(htmlString, attrFilters) {
  let filteredString = htmlString
  const { elements } = parseTags(htmlString)
  elements.forEach(({ nodeName, attributes }) => {
    attributes.forEach((attribute) => {
      if (attributeMatchConfig(nodeName, attribute, attrFilters)) {
        filteredString.replace(attribute.raw, '');
      }
    })
  });

  return filteredString;
}
```